### PR TITLE
Allow mod-tap hold action on one shot layer

### PR DESCRIPTION
**Pull request by kosorin**
## Description

This proposal allows mod-tap hold action on one shot layer. One shot layer is immediately disabled if `MT()` is pressed and therefore mod-tap hold is not possible.

### Example:

Following example currently produce a lower case character `x`.

```
 [L1] = LAYOUT(KC_X, OSL(L2))
 [L2] = LAYOUT(KC_A, LSFT_T(...))
```

- on layer 1
- `OSL(L2)` down
- `OSL(L2)` up
- on layer 2
- `LSFT_T(...)` down
- `KC_A` down
- `KC_A` up
- `LSFT_T(...)` up

With the proposed change the result is a upper case `A` and it will behave same as `KC_LSFT`.

I'm not aware if this covers all possible situations, but I've tested it for several months with no problems.

## Types of Changes

- [x] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: quantum/action.c [hunk 1]
```diff
@@ -356,7 +356,16 @@ void process_action(keyrecord_t *record, action_t action) {
...
+    if (is_oneshot_layer_active() 
+        && event.pressed 
+        && (action.kind.id == ACT_USAGE
+            || !(IS_MOD(action.key.code)
+#    ifndef NO_ACTION_TAPPING
+                || action.kind.id == ACT_LMODS_TAP
+                || action.kind.id == ACT_RMODS_TAP
+#    endif
+                )
+            )
```
Comment (drashna) [inline quantum/action.c:368]: Could you reformat that (`qmk format-c`)?

**Review by drashna — CHANGES_REQUESTED** [review]
Also, could you add to the unit tests to ensure that this behavior is properly tested?

**Comment by kosorin** [comment]
I added a few unit tests and fixed a small bug that I discovered thanks to the tests. I'm not sure that's enough. It certainly doesn't test all permutations of all Tap-Hold options.

**Comment by drashna** [comment]
> I added a few unit tests and fixed a small bug that I discovered thanks to the tests. I'm not sure that's enough. It certainly doesn't test all permutations of all Tap-Hold options.

Well, having them at all is a huge thing.  And they've already proven when even basic tests are definitely better than nothing, here. :)

At the very least, it helps with confidence in the code. And more robust testing can be added later.

**Review by drashna — APPROVED** [review]

**Review by KarlK90 — APPROVED** [review]
Thanks for fixing bugs and adding unit-tests in the progress! This is much appreciated. The changes LGTM.

**Pull request merged**